---
title: "report"
author: "Yihan"
date: "7/30/2021"
output:
  pdf_document: default
  html_document: default
---

```{r load libary}
library(DESeq2)
library(pheatmap)
library(ggplot2)
library(ggrepel)
library(tidyverse)
library(dplyr)
library(gplots)
library(RColorBrewer)
library(genefilter)
library(cowplot)
library(apeglm)
library(org.Mm.eg.db)
library(pathview) 
library(gageData)
library(gage) 
library(clusterProfiler) 
library(enrichplot)
library(vsn)
```

```{r import data}
data("go.sets.mm")
data("go.subs.mm")
data("kegg.sets.mm")
data("sigmet.idx.mm")
```

```{r functions}
# Read in sample sheet and append sample name as row labels
# Args: input.file: a .txt samplesheet
# Return: a data table
ReadTable <- function(input.file) {
  table <- read.table(input.file, header=T, sep="\t")
  rownames(table) <- table$SampleName
  return(table)
}

# Subsets sample sheet to only include specified tissue
# Args: table of sample sheet on samples
# Returns: a subsetted table
SetTissue <- function(table, tissue) {
  table <- subset(table, Tissue == tissue)
  return(table)
}
```

```{r read in the sample sheet}
# must read in as a 'text delimited' txt file
sampleTable <- ReadTable("data/Sample_sheet2.txt")
sampleTable

sampletable_med <- subset(sampleTable, Tissue == "med")
sampletable_med

sampletable_csc <- subset(sampleTable, Tissue == "csc")
sampletable_csc

# load in the count data to produce a count matrix
# directory is path to directory where the counts are stored (one per sample)
# design is how we wish to model the data: here we are measuring differences by genotype
ddsHTSeq <- DESeqDataSetFromHTSeqCount(sampleTable = sampleTable,
                                       directory = "data/4th_col",
                                       design = ~ 1)

dds_med <- DESeqDataSetFromHTSeqCount(sampleTable = sampletable_med,
                                       directory = "data/4th_col",
                                       design = ~ 1)

dds_csc <- DESeqDataSetFromHTSeqCount(sampleTable = sampletable_csc,
                                       directory = "data/4th_col",
                                       design = ~ 1)

```

```{r filter low counts and re-factor}
# pre-filter to exclude rows with few reads, reduce memory size of ddsHTSeq,
# dds_med, and dds_csc object
# increase speed of transformation and testing functions
ddsHTSeq_filtered <- ddsHTSeq[rowSums(counts(ddsHTSeq)) >= 10, ]
dds_med_filtered <- dds_med[rowSums(counts(dds_med)) >= 10,]
dds_csc_filtered <- dds_csc[rowSums(counts(dds_csc)) >= 10,]

# add a new "group" column to ddsHTSeq_filtered object by pasting 3 existing columns
ddsHTSeq_filtered$group <- factor(paste0(ddsHTSeq_filtered$Age, 
                                         ddsHTSeq_filtered$Genotype,
                                         ddsHTSeq_filtered$Tissue))
ddsHTSeq_filtered$group

# set Genotype, Age, Tissue as a factor and relevel 'WT' as baseline
ddsHTSeq_filtered$Genotype <- factor(ddsHTSeq_filtered$Genotype)
ddsHTSeq_filtered$Age <- factor(ddsHTSeq_filtered$Age)
ddsHTSeq_filtered$Tissue <- factor(ddsHTSeq_filtered$Tissue)
ddsHTSeq_filtered$Genotype <- relevel(ddsHTSeq_filtered$Genotype, ref = "WT")

# design function holds the formula of how the counts depend on the variables group
design(ddsHTSeq_filtered) <- ~ group

# add a new "group" column to dds_med_filtered object by pasting 2 existing columns
dds_med_filtered$group <- factor(paste0(dds_med_filtered$Age, 
                                         dds_med_filtered$Genotype))
dds_med_filtered$group

# set Genotype, Age, Tissue as a factor and relevel 'WT' as baseline
dds_med_filtered$Genotype <- factor(dds_med_filtered$Genotype)
dds_med_filtered$Age <- factor(dds_med_filtered$Age)
dds_med_filtered$Genotype <- relevel(dds_med_filtered$Genotype, ref = "WT")
design(dds_med_filtered) <- ~ group

# add a new "group" column to dds_csc_filtered object by pasting 2 existing columns
dds_csc_filtered$group <- factor(paste0(dds_csc_filtered$Age, 
                                         dds_csc_filtered$Genotype))

# set Genotype, Age, Tissue as a factor and relevel 'WT' as baseline
dds_csc_filtered$Genotype <- factor(dds_csc_filtered$Genotype)
dds_csc_filtered$Age <- factor(dds_csc_filtered$Age)
dds_csc_filtered$Genotype <- relevel(dds_csc_filtered$Genotype, ref = "WT")
design(dds_csc_filtered) <- ~ group
```

```{r transform matrix using variance stabilizing transformation}
# Quickly estimate dispersion trend and apply a variance stabilizing transformation
vsd_all <- vst(ddsHTSeq_filtered, blind = FALSE)
vsd_med <- vst(dds_med_filtered, blind = FALSE)
vsd_csc <- vst(dds_csc_filtered, blind = FALSE)
```

```{r fit statistical model}
dds <- DESeq(ddsHTSeq_filtered)
dds_med <- DESeq(dds_med_filtered)
dds_csc <- DESeq(dds_csc_filtered)
```

```{r variable genes med KI vs WT (9w)}
# results function extracts a results table with log2 fold changes, p values and 
# adjusted p values. More strict filtering to increase power is automatically applied via 
# independent filtering on the mean of normalized counts

# By genotype for med KI vs WT (9w)
res_9w_med_genotype <- results(dds, contrast=c("group","9wWTmed","9wKImed"), cooksCutoff=FALSE) 
res_9w_med_genotype[order(abs(res_9w_med_genotype$log2FoldChange), decreasing = TRUE),]
```

```{r variable genes csc KI vs WT (9w)}
# By genotype for csc KI vs WT (9w)
res_9w_csc_genotype <- results(dds, contrast=c("group","9wWTcsc","9wKIcsc"), cooksCutoff=FALSE) 
res_9w_csc_genotype[order(abs(res_9w_csc_genotype$log2FoldChange), decreasing = TRUE), ]
```

```{r variable genes for med KI vs WT (p)}
# By genotype for med KI vs WT (p)
res_p_med_genotype <- results(dds, contrast=c("group","pWTmed","pKImed"), cooksCutoff=FALSE) 
res_p_med_genotype[order(abs(res_p_med_genotype$log2FoldChange), decreasing = TRUE), ]
```

```{r variable genes csc KI vs WT (p)}
# By genotype for csc KI vs WT (p)
res_p_csc_genotype <- results(dds, contrast=c("group","pWTcsc","pKIcsc"), cooksCutoff=FALSE) 
res_p_csc_genotype[order(abs(res_p_csc_genotype$log2FoldChange), decreasing = TRUE), ]
```

```{r variable genes med KI vs WT (5w)}
# By genotype for med KI vs WT (5w)
res_5w_med_genotype <- results(dds, contrast=c("group","5wWTmed","5wKImed"), cooksCutoff=FALSE, independentFiltering = FALSE) 
res_5w_med_genotype[order(abs(res_5w_med_genotype$log2FoldChange), decreasing = TRUE), ]
```

```{r variable genes csc KI vs WT (5w)}
# By genotype for csc KI vs WT (5w)
res_5w_csc_genotype <- results(dds, contrast=c("group","5wWTcsc","5wKIcsc"), cooksCutoff=FALSE) 
res_5w_csc_genotype[order(abs(res_5w_csc_genotype$log2FoldChange), decreasing = TRUE), ]
```

```{r Heatmap of the count matrix}
# heatmap of med & csc
select_all <- order(rowMeans(counts(dds,normalized = TRUE)),
                decreasing=TRUE)[1:20]
df <- as.data.frame(colData(dds)[,c("Age","Genotype", "Tissue")])
pheatmap(assay(ntd)[select,], cluster_rows=FALSE, show_rownames=FALSE,
         cluster_cols=FALSE, annotation_col=df)

# heatmap of med 
select_med <- order(rowMeans(counts(dds_med,normalized = TRUE)),
                decreasing=TRUE)[1:20]
df <- as.data.frame(colData(dds_med)[,c("Age","Genotype")])
pheatmap(assay(ntd_med)[select_med,], cluster_rows=FALSE, show_rownames=FALSE,
         cluster_cols=FALSE, annotation_col=df)

# heatmap of csc
select_csc <- order(rowMeans(counts(dds_csc,normalized = TRUE)),
                decreasing=TRUE)[1:20]
df <- as.data.frame(colData(dds_csc)[,c("Age","Genotype")])
pheatmap(assay(ntd_csc)[select_csc,], cluster_rows=FALSE, show_rownames=FALSE,
         cluster_cols=FALSE, annotation_col=df)
```

```{r volcano plots_2_9w_med_genotype}
library(EnhancedVolcano)
volcano_9w <- EnhancedVolcano(res_9w_med_genotype,
    lab = rownames(res_9w_med_genotype),
    title = '9w med by genotype',
    subtitle = paste0('p-value cutoff (red line) drawn ',
      'at equivalent of adjusted p=0.01'),
    x = 'log2FoldChange',
    xlim = c(-6, 6),
    ylim = c(0, 40),
    y = 'pvalue',
    pCutoff = 0.01,
    FCcutoff = 1.5,
    pointSize = 3.0,
    labSize = 6.0,
    colAlpha = 1,
    cutoffLineType = 'solid',
    cutoffLineCol = 'red2',
    cutoffLineWidth = 2.5,
    hline = c(10e-20,
      10e-20 * 10e-30,
      10e-20 * 10e-60,
      10e-20 * 10e-90),
    hlineCol = c('black', 'black', 'black', 'black'),
    hlineType = c('longdash', 'longdash', 'dotdash', 'dotdash'),
    hlineWidth = c(0.4, 0.4, 0.8, 0.8),
    gridlines.major = FALSE,
    gridlines.minor = FALSE) + 
  coord_flip()
volcano_9w
```

```{r sample-to-sample distances}
library("RColorBrewer")
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)

# vsd of all
sampleDists <- dist(t(assay(vsd)))
sampleDistMatrix <- as.matrix(sampleDists)
rownames(sampleDistMatrix) <- paste(vsd$Age, vsd$Genotype, sep="-")
colnames(sampleDistMatrix) <- NULL
pheatmap(sampleDistMatrix,
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists,
         col=colors)

# vsd of med
sampleDists_med <- dist(t(assay(vsd_med)))
sampleDistMatrix_med <- as.matrix(sampleDists_med)
rownames(sampleDistMatrix_med) <- paste(vsd_med$Age, vsd_med$Genotype, sep="-")
colnames(sampleDistMatrix_med) <- NULL
pheatmap(sampleDistMatrix_med,
         clustering_distance_rows=sampleDists_med,
         clustering_distance_cols=sampleDists_med,
         col=colors)

# vsd of csc
sampleDists_csc <- dist(t(assay(vsd_csc)))
sampleDistMatrix_csc <- as.matrix(sampleDists_csc)
rownames(sampleDistMatrix_csc) <- paste(vsd_csc$Age, vsd_csc$Genotype, sep="-")
colnames(sampleDistMatrix_csc) <- NULL
pheatmap(sampleDistMatrix_csc,
         clustering_distance_rows=sampleDists_csc,
         clustering_distance_cols=sampleDists_csc,
         col=colors)
```

```{r PCA plot os samples}
# customized PCA Plot for med & csc
pcaData <- plotPCA(vsd, intgroup=c("Age", "Genotype"), returnData=TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar"))
ggplot(pcaData, aes(PC1, PC2, color = Genotype, shape = Age)) +
  geom_point(size=3) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  coord_fixed()


# customized PCA Plot for med
pcaData_med <- plotPCA(vsd_med, intgroup=c("Age", "Genotype"), returnData=TRUE)
percentVar <- round(100 * attr(pcaData_med, "percentVar"))
ggplot(pcaData_med, aes(PC1, PC2, color = Genotype, shape = Age)) +
  geom_point(size=3) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  coord_fixed() +
  labs(title = "PCA Plot for Medulla")

# customized PCA Plot for csc
pcaData_csc <- plotPCA(vsd_csc, intgroup=c("Age", "Genotype"), returnData=TRUE)
percentVar <- round(100 * attr(pcaData_csc, "percentVar"))
ggplot(pcaData_csc, aes(PC1, PC2, color = Genotype, shape = Age)) +
  geom_point(size=3) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  coord_fixed() + 
  labs(title = "PCA Plot for Cervical spinal cord")
```